flow/directives.py

```diff
@@ -0,0 +1,315 @@
+
+    def __getitem__(self, key):
+        try:
+            value = self._defined_directives[key]
+            return self._directive_definitions[key].finalize(value, self)
+        except KeyError:
+            try:
+                return self._user_directives[key]
+            except KeyError:
+                raise KeyError("{} not in directives.".format(key))
+
+    def __setitem__(self, key, value):
+        if key in self._directive_definitions:
+            self._set_defined_directive(key, value)
+        else:
+            self._user_directives[key] = value
+
+    def __delitem__(self, key):
+        if key in self._directive_definitions:
+            self._defined_directives[key] = self._directive_definitions[key].default
+        else:
+            del self._user_directives[key]
+
+    def __iter__(self):
+        yield from self._defined_directives
+        yield from self._user_directives
+
+    def __len__(self):
+        return len(self._defined_directives) + len(self._user_directives)
+
+    def __str__(self):
+        return str(dict(self))
+
+    def __repr__(self):
+        return "Directives({})".format(str(self))
+
+    def update(self, other, aggregate=False, job=None, parallel=False):
+        if aggregate:
+            self._aggregate(other, job=job, parallel=parallel)
+        else:
+            super().update(other)
+
+    def evaluate(self, job):
+        for key, value in self.items():
+            self[key] = _evaluate(value, job)
+
+    def _aggregate(self, other, job=None, parallel=False):
+        self.evaluate(job)
+        other.evaluate(job)
+        agg_func_attr = "parallel" if parallel else "serial"
+        for name in self._defined_directives:
+            agg_func = getattr(self._directive_definitions[name], agg_func_attr)
+            dft = self._directive_definitions[name].default
+            other_directive = other.get(name, dft)
+            directive = self[name]
+            if other_directive is None:
+                continue
+            else:
+                self._defined_directives[name] = agg_func(directive,
+                                                          other_directive)
+
+
+def _evaluate(value, job=None):
+    if callable(value):
+        if job is None:
+            raise RuntimeError("job not specified when evaluating directive.")
+        else:
+            return value(job)
+    else:
+        return value
+
+
+class OnlyType:
+    def __init__(self, type_, preprocess=None, postprocess=None):
+        def identity(v):
+            return v
+
+        self.type = type_
+        self.preprocess = identity if preprocess is None else preprocess
+        self.postprocess = identity if postprocess is None else postprocess
+
+    def __call__(self, v):
+        return self.postprocess(self._validate(self.preprocess(v)))
+
+    def _validate(self, v):
+        if isinstance(v, self.type):
+            return v
+        else:
+            try:
+                return self.type(v)
```

Comment:
@b-butler  If we pass in decimal when the required value is a validation, it converts that decimal into an integer. (This is not the expected behaviour, right?)
If not then we should raise error straightaway.

Reply:
When I pass in `None` which is of type `NoneType` to `Executable` then it creates string `None` and raises no error.

Reply:
Some of these decisions are difficult. In a way, we don't want a user to do something like `value = 4 / 2` and raise an error (the output of division is always a float in Python. In this case, I think we allow it. We can be more strict by checking how close it is to an integer, but I do not feel that is necessary for now.

Reply:
Regarding the `None`, conversion. We should have an allow none flag in the class which just returns `None` if the value passed in is `None`.

Reply:
okay, noted